### PR TITLE
Archive summary

### PR DESCRIPTION
**Pull request by zschira**
This PR generates a `RunSummary` after every archiver run, which contains key details such as the number of files that have changed, the results of all validation tests, and metadata about the new version and previous version. The summary can be output as a JSON file, which will be attached to any runs of the `run-archiver` github action.

### CODE REVIEW

File: .github/workflows/run-archiver.yml [hunk 1]
```diff
@@ -68,7 +68,13 @@ jobs:
           # ZENODO_TOKEN_UPLOAD: ${{ secrets.ZENODO_TOKEN_UPLOAD }}
           # ZENODO_TOKEN_PUBLISH: ${{ secrets.ZENODO_TOKEN_PUBLISH }}
         run: |
-          conda run -n pudl-cataloger pudl_archiver --sandbox --datasets ${{ matrix.dataset }}
+          conda run -n pudl-cataloger pudl_archiver --sandbox --datasets ${{ matrix.dataset }} --sumary-file ${{ matrix.dataset }}_run_summary.json
```
Comment (jdangerx) [inline .github/workflows/run-archiver.yml:71]: nit: this flag should probably be called `--summary-file`

File: .github/workflows/run-archiver.yml [hunk 2]
```diff
@@ -68,7 +68,13 @@ jobs:
...
+
+      - name: Archive code coverage results
+        uses: actions/upload-artifact@v3
+        with:
+          name: $${{ matrix.dataset }}-run-summary
+          path: ${{ matrix.dataset }}_run_summary.json
```
Comment (jdangerx) [inline .github/workflows/run-archiver.yml:77]: For now, we'd only access these through the Actions UI, right?

File: src/pudl_archiver/__init__.py [hunk 3]
```diff
@@ -79,3 +81,13 @@ async def archive_datasets(
...
+
+    if summary_file is not None:
+        run_summaries = [result.dict() for _, result in results]
+
+        with open(summary_file, "w") as f:
+            json.dump(run_summaries, f, indent=2)
+
+    validation_results = [result.success for _, result in results]
+    if not all(validation_results):
+        raise RuntimeError("Error: archive validation tests failed.")
```
Comment (jdangerx) [inline src/pudl_archiver/__init__.py:93]: Is this your imagined workflow?:

1. see this
2. shoot, need to go look at the run summary file

File: tests/unit/archive_validate_test.py [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+                    hash="hash1",
+                ),
+            },
+            [],
+        ),
+    ],
+)
+def test_process_resource_diffs(baseline_resources, new_resources, diffs):
+    """Test resource diffs."""
+    test_diffs = validate._process_resource_diffs(baseline_resources, new_resources)
```
Comment (jdangerx) [inline tests/unit/archive_validate_test.py:265]: Typically I'd avoid testing the private function and test the public interface instead - here, that'd be `RunSummary.create_summary`. 

It's 6PM on a Friday so no code sample, but I'll crank on one on Monday morning so you have something to actually compare things to.

File: tests/unit/archive_validate_test.py [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+        ),
+    ],
+)
+def test_process_partition_diffs(baseline_partitions, new_partitions, diffs):
+    """Test partition diffs."""
+    test_diffs = validate._process_partition_diffs(baseline_partitions, new_partitions)
+    assert test_diffs == diffs
+
+
+@pytest.mark.parametrize(
```
Comment (jdangerx) [inline tests/unit/archive_validate_test.py:58]: Good instinct to parametrize, though I personally still find this data kind of hard to follow just because there's so many lines of it. Seems like there's a lot of repetition with small tweaks in the data - maybe consider a test data factory that churns out similar-looking data? I'll mess around a bit on Monday and see what I come up with.

File: tests/unit/archive_base_test.py [hunk 6]
```diff
@@ -235,7 +235,10 @@ async def test_get_hyperlinks(docname, pattern, links, request, html_docs):
 def test_validate_archive(datapackage, test_results, success):
```
Comment (jdangerx) [inline tests/unit/archive_base_test.py:235]: nit: probably we should change the name of this test if the method under test is changing.

File: src/pudl_archiver/orchestrator.py [hunk 7]
```diff
@@ -191,28 +190,29 @@ async def run(self):
...
-            return self.new_deposition
+            return Unchanged(dataset_name=self.data_source_id, reason="Dry run.")
 
         await self._apply_changes()
         self.new_deposition = await self.depositor.get_record(self.new_deposition.id_)
         new_datapackage, old_datapackage = await self._update_datapackage(resources)
         await self._apply_changes()
 
         if self.changed:
-            validation_success = self.downloader.validate_archive(
```
Comment (jdangerx) [inline src/pudl_archiver/orchestrator.py:202]: The first structure that jumps to mind is

```
orchestrator
    |--> downloader
    |--> validator/summarizer
    |--> uploader
```

Where the three "thing-doers" are siblings and the orchestrator tells them what to do.

What were you considering when thinking about where to put the validation/summary stuff?

File: tests/integration/zenodo_test.py [hunk 8]
```diff
@@ -226,16 +246,31 @@ async def identity(x):
         )
     )
 
-    v2 = await orchestrator.run()
-    v2_refreshed = await orchestrator.depositor.get_record(v2.id_)
+    # Should fail due to deleted file
+    v2_summary = await orchestrator.run()
```
Comment (jdangerx) [inline tests/integration/zenodo_test.py:250]: Is the idea that this will remove a file, and `check_missing_files` says "if there was a file before, and now that file is gone, SCREAM"?

**Review by jdangerx — CHANGES_REQUESTED** [review]
This looks like it will work, and be super useful eventually! I've some readability / structure suggestions to discuss before we merge, and will play around with the actual code on Monday morning so I'm not just idly speculating on what "could" be better.

File: tests/unit/archive_validate_test.py (hunk 4 above)
Comment (jdangerx) [reply tests/unit/archive_validate_test.py:265]:
```diff
diff --git a/tests/unit/archive_validate_test.py b/tests/unit/archive_validate_test.py
index 46eb31a..7b332b9 100644
--- a/tests/unit/archive_validate_test.py
+++ b/tests/unit/archive_validate_test.py
@@ -59,8 +59,8 @@ def test_process_partition_diffs(baseline_partitions, new_partitions, diffs):
     "baseline_resources,new_resources,diffs",
     [
         (
-            {
-                "resource0": Resource(
+            [
+                Resource(
                     name="resource0",
                     path="https://www.fake.link",
                     remote_url="https://www.fake.link",
@@ -71,7 +71,7 @@ def test_process_partition_diffs(baseline_partitions, new_partitions, diffs):
                     bytes=10,
                     hash="hash0",
                 ),
-                "resource1": Resource(
+                Resource(
                     name="resource1",
                     path="https://www.fake.link",
                     remote_url="https://www.fake.link",
@@ -82,9 +82,9 @@ def test_process_partition_diffs(baseline_partitions, new_partitions, diffs):
                     bytes=20,
                     hash="hash1",
                 ),
-            },
-            {
-                "resource0": Resource(
+            ],
+            [
+                Resource(
                     name="resource0",
                     path="https://www.fake.link",
                     remote_url="https://www.fake.link",
@@ -95,7 +95,7 @@ def test_process_partition_diffs(baseline_partitions, new_partitions, diffs):
                     bytes=10,
                     hash="hash0",
                 ),
-                "resource1": Resource(
+                Resource(
                     name="resource1",
                     path="https://www.fake.link",
                     remote_url="https://www.fake.link",
@@ -106,7 +106,7 @@ def test_process_partition_diffs(baseline_partitions, new_partitions, diffs):
                     bytes=40,
                     hash="hash1_changed",
                 ),
-            },
+            ],
             [
                 validate.FileDiff(
                     name="resource1",
@@ -116,8 +116,8 @@ def test_process_partition_diffs(baseline_partitions, new_partitions, diffs):
             ],
         ),
         (
-            {
-                "resource0": Resource(
+            [
+                Resource(
                     name="resource0",
                     path="https://www.fake.link",
                     remote_url="https://www.fake.link",
@@ -128,7 +128,7 @@ def test_process_partition_diffs(baseline_partitions, new_partitions, diffs):
                     bytes=10,
                     hash="hash0",
                 ),
-                "resource1": Resource(
+                Resource(
                     name="resource1",
                     path="https://www.fake.link",
                     remote_url="https://www.fake.link",
@@ -139,9 +139,9 @@ def test_process_partition_diffs(baseline_partitions, new_partitions, diffs):
                     bytes=20,
                     hash="hash1",
                 ),
-            },
-            {
-                "resource0": Resource(
+            ],
+            [
+                Resource(
                     name="resource0",
                     path="https://www.fake.link",
                     remote_url="https://www.fake.link",
@@ -152,7 +152,7 @@ def test_process_partition_diffs(baseline_partitions, new_partitions, diffs):
                     bytes=10,
                     hash="hash0",
                 ),
-            },
+            ],
             [
                 validate.FileDiff(
                     name="resource1",
@@ -162,8 +162,8 @@ def test_process_partition_diffs(baseline_partitions, new_partitions, diffs):
             ],
         ),
         (
-            {
-                "resource0": Resource(
+            [
+                Resource(
                     name="resource0",
                     path="https://www.fake.link",
                     remote_url="https://www.fake.link",
@@ -174,9 +174,9 @@ def test_process_partition_diffs(baseline_partitions, new_partitions, diffs):
                     bytes=10,
                     hash="hash0",
                 ),
-            },
-            {
-                "resource0": Resource(
+            ],
+            [
+                Resource(
                     name="resource0",
                     path="https://www.fake.link",
                     remote_url="https://www.fake.link",
@@ -187,7 +187,7 @@ def test_process_partition_diffs(baseline_partitions, new_partitions, diffs):
                     bytes=10,
                     hash="hash0",
                 ),
-                "resource1": Resource(
+                Resource(
                     name="resource1",
                     path="https://www.fake.link",
                     remote_url="https://www.fake.link",
@@ -198,7 +198,7 @@ def test_process_partition_diffs(baseline_partitions, new_partitions, diffs):
                     bytes=20,
                     hash="hash1_created",
                 ),
-            },
+            ],
             [
                 validate.FileDiff(
                     name="resource1",
@@ -208,8 +208,8 @@ def test_process_partition_diffs(baseline_partitions, new_partitions, diffs):
             ],
         ),
         (
-            {
-                "resource0": Resource(
+            [
+                Resource(
                     name="resource0",
                     path="https://www.fake.link",
                     remote_url="https://www.fake.link",
@@ -220,7 +220,7 @@ def test_process_partition_diffs(baseline_partitions, new_partitions, diffs):
                     bytes=10,
                     hash="hash0",
                 ),
-                "resource1": Resource(
+                Resource(
                     name="resource1",
                     path="https://www.fake.link",
                     remote_url="https://www.fake.link",
@@ -231,9 +231,9 @@ def test_process_partition_diffs(baseline_partitions, new_partitions, diffs):
                     bytes=20,
                     hash="hash1",
                 ),
-            },
-            {
-                "resource0": Resource(
+            ],
+            [
+                Resource(
                     name="resource0",
                     path="https://www.fake.link",
                     remote_url="https://www.fake.link",
@@ -244,7 +244,7 @@ def test_process_partition_diffs(baseline_partitions, new_partitions, diffs):
                     bytes=10,
                     hash="hash0",
                 ),
-                "resource1": Resource(
+                Resource(
                     name="resource1",
                     path="https://www.fake.link",
                     remote_url="https://www.fake.link",
@@ -255,12 +255,24 @@ def test_process_partition_diffs(baseline_partitions, new_partitions, diffs):
                     bytes=20,
                     hash="hash1",
                 ),
-            },
+            ],
             [],
         ),
     ],
 )
-def test_process_resource_diffs(baseline_resources, new_resources, diffs):
+def test_process_resource_diffs(baseline_resources, new_resources, diffs, mocker):
     """Test resource diffs."""
-    test_diffs = validate._process_resource_diffs(baseline_resources, new_resources)
+    baseline = mocker.MagicMock(
+        resources=baseline_resources, version="0.0.1", created="2023-01-01"
+    )
+    new = mocker.MagicMock(
+        resources=new_resources, version="0.0.2", created="2023-01-02"
+    )
+    summary = validate.RunSummary.create_summary(
+        name="test package",
+        baseline_datapackage=baseline,
+        new_datapackage=new,
+        validation_tests=[],
+    )
+    test_diffs = summary.file_changes
     assert test_diffs == diffs
```

Comment (jdangerx) [reply tests/unit/archive_validate_test.py:265]: This leaves us freer to mess with the implementation details later down the line.

File: tests/unit/archive_validate_test.py (hunk 5 above)
Comment (jdangerx) [reply tests/unit/archive_validate_test.py:58]: Here's an example of what such a factory might look like, and how it makes it easier to add more test cases:

```diff
diff --git a/tests/unit/archive_validate_test.py b/tests/unit/archive_validate_test.py
index 46eb31a..1b7d384 100644
--- a/tests/unit/archive_validate_test.py
+++ b/tests/unit/archive_validate_test.py
@@ -55,58 +55,30 @@ def test_process_partition_diffs(baseline_partitions, new_partitions, diffs):
     assert test_diffs == diffs
 
 
+def fake_resource(num=0, **kwargs):
+    params = {
+        "name": f"resource{num}",
+        "path": "https://www.fake.link",
+        "remote_url": "https://www.fake.link",
+        "title": f"Resource {num}",
+        "parts": {},
+        "mediatype": "zip",
+        "format": "format",
+        "bytes": 10,
+        "hash": f"hash{num}",
+    } | kwargs
+    return Resource(**params)
+
+
 @pytest.mark.parametrize(
     "baseline_resources,new_resources,diffs",
     [
         (
-            {
-                "resource0": Resource(
-                    name="resource0",
-                    path="https://www.fake.link",
-                    remote_url="https://www.fake.link",
-                    title="Resource 0",
-                    parts={},
-                    mediatype="zip",
-                    format="format",
-                    bytes=10,
-                    hash="hash0",
-                ),
-                "resource1": Resource(
-                    name="resource1",
-                    path="https://www.fake.link",
-                    remote_url="https://www.fake.link",
-                    title="Resource 1",
-                    parts={},
-                    mediatype="zip",
-                    format="format",
-                    bytes=20,
-                    hash="hash1",
-                ),
-            },
-            {
-                "resource0": Resource(
-                    name="resource0",
-                    path="https://www.fake.link",
-                    remote_url="https://www.fake.link",
-                    title="Resource 0",
-                    parts={},
-                    mediatype="zip",
-                    format="format",
-                    bytes=10,
-                    hash="hash0",
-                ),
-                "resource1": Resource(
-                    name="resource1",
-                    path="https://www.fake.link",
-                    remote_url="https://www.fake.link",
-                    title="Resource 1",
-                    parts={},
-                    mediatype="zip",
-                    format="format",
-                    bytes=40,
-                    hash="hash1_changed",
-                ),
-            },
+            [
+                fake_resource(0),
+                fake_resource(1, bytes=20),
+            ],
+            [fake_resource(0), fake_resource(1, bytes=40, hash="hash1_changed")],
             [
                 validate.FileDiff(
                     name="resource1",
@@ -116,43 +88,11 @@ def test_process_partition_diffs(baseline_partitions, new_partitions, diffs):
             ],
         ),
         (
-            {
-                "resource0": Resource(
-                    name="resource0",
-                    path="https://www.fake.link",
-                    remote_url="https://www.fake.link",
-                    title="Resource 0",
-                    parts={},
-                    mediatype="zip",
-                    format="format",
-                    bytes=10,
-                    hash="hash0",
-                ),
-                "resource1": Resource(
-                    name="resource1",
-                    path="https://www.fake.link",
-                    remote_url="https://www.fake.link",
-                    title="Resource 1",
-                    parts={},
-                    mediatype="zip",
-                    format="format",
-                    bytes=20,
-                    hash="hash1",
-                ),
-            },
-            {
-                "resource0": Resource(
-                    name="resource0",
-                    path="https://www.fake.link",
-                    remote_url="https://www.fake.link",
-                    title="Resource 0",
-                    parts={},
-                    mediatype="zip",
-                    format="format",
-                    bytes=10,
-                    hash="hash0",
-                ),
-            },
+            [
+                fake_resource(0),
+                fake_resource(1, bytes=20),
+            ],
+            [fake_resource(0)],
             [
                 validate.FileDiff(
                     name="resource1",
@@ -162,43 +102,8 @@ def test_process_partition_diffs(baseline_partitions, new_partitions, diffs):
             ],
         ),
         (
-            {
-                "resource0": Resource(
-                    name="resource0",
-                    path="https://www.fake.link",
-                    remote_url="https://www.fake.link",
-                    title="Resource 0",
-                    parts={},
-                    mediatype="zip",
-                    format="format",
-                    bytes=10,
-                    hash="hash0",
-                ),
-            },
-            {
-                "resource0": Resource(
-                    name="resource0",
-                    path="https://www.fake.link",
-                    remote_url="https://www.fake.link",
-                    title="Resource 0",
-                    parts={},
-                    mediatype="zip",
-                    format="format",
-                    bytes=10,
-                    hash="hash0",
-                ),
-                "resource1": Resource(
-                    name="resource1",
-                    path="https://www.fake.link",
-                    remote_url="https://www.fake.link",
-                    title="Resource 1",
-                    parts={},
-                    mediatype="zip",
-                    format="format",
-                    bytes=20,
-                    hash="hash1_created",
-                ),
-            },
+            [fake_resource(0)],
+            [fake_resource(0), fake_resource(1, bytes=20)],
             [
                 validate.FileDiff(
                     name="resource1",
@@ -208,59 +113,30 @@ def test_process_partition_diffs(baseline_partitions, new_partitions, diffs):
             ],
         ),
         (
-            {
-                "resource0": Resource(
-                    name="resource0",
-                    path="https://www.fake.link",
-                    remote_url="https://www.fake.link",
-                    title="Resource 0",
-                    parts={},
-                    mediatype="zip",
-                    format="format",
-                    bytes=10,
-                    hash="hash0",
-                ),
-                "resource1": Resource(
-                    name="resource1",
-                    path="https://www.fake.link",
-                    remote_url="https://www.fake.link",
-                    title="Resource 1",
-                    parts={},
-                    mediatype="zip",
-                    format="format",
-                    bytes=20,
-                    hash="hash1",
-                ),
-            },
-            {
-                "resource0": Resource(
-                    name="resource0",
-                    path="https://www.fake.link",
-                    remote_url="https://www.fake.link",
-                    title="Resource 0",
-                    parts={},
-                    mediatype="zip",
-                    format="format",
-                    bytes=10,
-                    hash="hash0",
-                ),
-                "resource1": Resource(
-                    name="resource1",
-                    path="https://www.fake.link",
-                    remote_url="https://www.fake.link",
-                    title="Resource 1",
-                    parts={},
-                    mediatype="zip",
-                    format="format",
-                    bytes=20,
-                    hash="hash1",
-                ),
-            },
+            [fake_resource(0), fake_resource(1, bytes=20)],
+            [fake_resource(0), fake_resource(1, bytes=20)],
             [],
         ),
+        (
+            [fake_resource(0)],
+            [fake_resource(0, hash="hash0_changed")],
+            [validate.FileDiff(name="resource0", diff_type="UPDATE", size_diff=0)],
+        ),
     ],
 )
def test_process_resource_diffs(baseline_resources, new_resources, diffs):
     """Test resource diffs."""
     test_diffs = validate._process_resource_diffs(baseline_resources, new_resources)
     assert test_diffs == diffs
```

File: .github/workflows/run-archiver.yml (hunk 2 above)
Comment (zschira) [reply .github/workflows/run-archiver.yml:77]: Yeah right now that's the case. I've thought about adding a json file to the actual depositions, so there will be a history right in the deposition. Do you think this would be worth doing?

File: src/pudl_archiver/__init__.py (hunk 3 above)
Comment (zschira) [reply src/pudl_archiver/__init__.py:93]: Yeah I thought about printing out any failed validation tests here, but decided it was sufficient to just go to the file if there's a failure. I'm definitely open to changing this though

File: tests/integration/zenodo_test.py (hunk 8 above)
Comment (zschira) [reply tests/integration/zenodo_test.py:250]: Yeah the `check_missing_files` test will be enabled by default since that would generally indicate something like a missing year, so this will first check that it failed, then switch the test off and check that it succeeds the second time.

File: .github/workflows/run-archiver.yml (hunk 2 above)
Comment (jdangerx) [reply .github/workflows/run-archiver.yml:77]: Sometime down the line, I guess that could be useful. But this seems fine for now.

File: src/pudl_archiver/orchestrator.py (hunk 7 above)
Comment (zschira) [reply src/pudl_archiver/orchestrator.py:202]: The reason I put the validator/summarizer stuff in the downloader is so each dataset can override the `dataset_validate_archive` method to add one off validation tests. One other pattern I could see is adding a decorator for adding validation tests, so tests could still be defined in the same module as a datasets archiver without needing them to be in the class itself. Something like:

```py
class Eia860Archiver(AbstractDatasetArchiver):
    # Do downloading stuff


@validation_test(dataset_name="eia860")
def test_something():
    return True
```

Do you think this pattern would make more sense? It would allow us to keep the downloading/validating/uploading separate

Comment (jdangerx) [reply src/pudl_archiver/orchestrator.py:202]: Ah, so basically if we have the downloader & the validator/summarizer be different classes, we end up with two dataset-specific classes per dataset. Which is annoying for sure. I think I'd rather keep everything in the same class than have a bunch of dangling validation tests in the module, that separation doesn't seem useful enough to justify the sort of strange pattern.

File: .github/workflows/run-archiver.yml (hunk 1 above)
Comment (zschira) [reply .github/workflows/run-archiver.yml]: Fixed

File: tests/unit/archive_validate_test.py (hunk 4 above)
Comment (zschira) [reply tests/unit/archive_validate_test.py]: Changed, thanks

File: tests/unit/archive_validate_test.py (hunk 5 above)
Comment (zschira) [reply tests/unit/archive_validate_test.py:73]: Good call, much cleaner now

File: tests/unit/archive_base_test.py (hunk 6 above)
Comment (zschira) [reply tests/unit/archive_base_test.py]: Fixed

File: src/pudl_archiver/orchestrator.py (hunk 7 above)
Comment (zschira) [reply src/pudl_archiver/orchestrator.py:202]: Yeah ideally we'll get to the place where if you subclass the `AbstractDatasetArchiver` and implement a couple of basic methods, you'll end up with new archives being auto-generated  and not need to think about anything else

**Review by jdangerx — APPROVED** [review]
:shipit:
